### PR TITLE
feat: configurable scopes with semantic matching

### PR DESCRIPTION
**Pull request by philocalyst**
In this PR I added scope configuration, which allows you to assign rules for the scopes to be added when you use koji, under specific rules. 

The ruleset is simple:
- Either you regex match over the files in the change.
- Or you more creatively setup an [ast-grep](https://ast-grep.github.io/) match, where you can configure based on the patterns in your code (example, a test-case change.)

### CODE REVIEW

**Comment by cococonscious** [comment]
Besides the conflicts, make sure to consider my comment in the previous PR: https://github.com/cococonscious/koji/pull/161#discussion_r2606507939 and imo it should also be possible to configure to force the use of the configured scopes or not (and with that if an empty scope is allowed or not)

File: src/lib/questions.rs [hunk 1]
```diff
@@ -148,6 +163,21 @@ impl Autocomplete for ScopeAutocompleter {
...
+    if config.force_scope && !config.commit_scopes.is_empty() {
+        let scope_values: Vec<String> = config.commit_scopes.keys().cloned().collect();
+
+        let prompt = Select::new("What's the scope of this change?", scope_values)
+            .with_render_config(get_render_config());
+
+        let result = if config.allow_empty_scope {
+            prompt.prompt_skippable()?
+        } else {
+            Some(prompt.prompt()?)
```
Comment (cococonscious) [inline src/lib/questions.rs]: Pressing \<Esc> cancels the prompt with a generic error message, please find a way to improve it (for example "Error: Scope cannot be empty"

**Review by cococonscious — CHANGES_REQUESTED** [review]
Looks good besides the review comment, also, I have two other points to consider:

- The `force_scope` config key naming is a bit misleading, I think a more accurate name would be `force_config_scopes`
- We should show the configured scope description in the selection prompt, not only because it's otherwise obsolete but also to differentiate from the detected scopes

**Comment by philocalyst** [comment]
I'll address #2, but #1 I wrote because it's namespaced under Config. So Config.force_config_scopes feels redundant.

You'll see updates later today.

▰▰▰▰▰
Miles Wirth 🙃

**Comment by cococonscious** [comment]
I understand your point about the scoping of the config property, and it does make sense, but it's a bit confusing because "force scope" could be misunderstood as "you must specify a scope" (which the other variable is for so it is understandable, yes, but it does require an extra thinking step).

Thanks for the changes. I did find a bug though, when setting the following: `force_scope = false` and `allow_empty_scope = false`, hitting \<Esc> will just continue instead of throwing like when `force_scope = true`.

Also, I find it hard to justify adding another error handling dependency, could we not either replace everything with `thiserror` or implement your changes in `anyhow`?

Lastly, the descriptions of the commit scopes defined in the config are still not visible.


Thanks,
Finley

**Comment by philocalyst** [comment]
The error handling thing: I think if we want to move to stronger error handling, the answer is 100% thiserror!! It's meant for structured error cases. The other thing we could consider is [Ariadne](https://lib.rs/crates/ariadne) or [Miette](https://lib.rs/crates/miette), which would place us in the top 1% for these diagnostics, but would require a larger refactor.

I could start a separate PR that adds that to merge before this one so consistency stays. 

I also was thinking that globbing would pair nicely with this scope configuration, as a lot of scopes are bound purely to specific sets of files from within a codebase. I have a scope for my flake.nix for example. That way users could manage LOTS of scopes without it polluting the main dialog, which would be reserved for the more universal options.

▰▰▰▰▰
Miles Wirth 🙃

**Comment by cococonscious** [comment]
> I think if we want to move to stronger error handling, the answer is 100% thiserror!! ... I could start a separate PR that adds that to merge before this one so consistency stays. 

Yes, please! That would be great.

> The other thing we could consider is [Ariadne](https://lib.rs/crates/ariadne) or [Miette](https://lib.rs/crates/miette), ...

While they look great, I think they're overkill for the simple errors that are common in koji, unless you disagree...

> I also was thinking that globbing would pair nicely with this scope configuration, ...

Sorry, I'm a bit lost on what exactly you mean by this, could you give me a concrete example?

**Comment by philocalyst** [comment]
Okay, I'll look into the error handling sometime soon!

They are basically overkill, just thinking if the config becomes a bigger interface it would help, and we can fallback to the default case elsewhere? But definitely to think about LATER.

Yes, something like

flakes = "/flake.nix"
core = "/crates/core/**/*.rs"
build = ["/build.rs", "/justfile"]

So if changes impact a match, the scope is pre-assigned.

Which assigns these things. On second thought, a Regex or FancyRegex implementation could make more sense, as I see the deliniation getting advanced, and arrays feel like a clunky answer.

▰▰▰▰▰
Miles Wirth 🙃

**Comment by philocalyst** [comment]
Following up @cococonscious

**Comment by cococonscious** [comment]
Sorry @philocalyst for the delayed answer.

I think that's a great idea, especially if we use a regex implementation instead of arrays. 

How would we handle precedence? In your example, if both "flake.nix" and "build.rs" are modified, would "flakes" or "build" be the chosen scope (if any)?

**Comment by philocalyst** [comment]
The thing about working on a tool for commits is that you develop standards just by making ergonomic decisions like this. I personally lean towards the atomic commit model, so my answer is "neither" as I believe any situation where you need to make that cross cutting change should be grouped into a PR. 

If that's something you agree with, that can be the stance, but long-term we may need to adopt a particular committing philosphy as to not grow too complex.

**Comment by philocalyst** [comment]
Regex is also a great idea, in that we could use the implementation of [ast-grep](https://ast-grep.github.io/) to both match on filenames and on code structure. Then you could do things like "if this modified function holds the test annotation, mark its scope as test".

Which is very exciting and a fair bit more technical, but I believe personally do-able. I volunteer it really because it differentiates Koji and uses ast-grep in a creative way (I find it difficult to use practically otherwise, despite how capable it is)

**Comment by philocalyst** [comment]
We could use ast_greps [YAML input](https://ast-grep.github.io/reference/yaml.html) (Appropriated to TOML of course), which I found when researching -- it makes incorporating this feel very "right" with the current config structure :)))

**Comment by cococonscious** [comment]
I agree with you on the atomic commits, although I suspect that in reality most people using koji will not fully adopt that philosophy, so your (great) idea about using ast-grep might not work as expected for everyone. I think it would be an awesome addition, but it should be gated behind a feature flag for now, until we get some experience and user reports about how they use and like it. The regex config approach can be in the default feature-set though. If you're willing to, please do implement both 🥇 
I'll try to review it faster than I've been replying in this thread, promise :)

**Comment by philocalyst** [comment]
Feeling a lot better about this! @cococonscious

**Comment by philocalyst** [comment]
@cococonscious Following up

File: src/lib/questions.rs [hunk 2]
```diff
@@ -147,20 +185,91 @@ impl Autocomplete for ScopeAutocompleter {
...
+    scope_values
+}
+
+fn prompt_scope(config: &Config, scope_matches: &ScopeMatches) -> Result<Option<String>> {
+    if config.force_scope && !config.commit_scopes.is_empty() {
+        if let Some(scope) = scope_matches.suggested() {
+            return Ok(Some(scope));
+        }
+
+        let scope_values = get_force_scope_values(config, scope_matches);
```
Comment (cococonscious) [inline src/lib/questions.rs]: The description of the configured scopes should also be shown when force_scope (or whichever name the config variable will have) is set to false. It will make it easy for the users to differentiate between configured and parsed scopes.

File: Cargo.toml [hunk 3]
```diff
@@ -26,19 +26,24 @@ cocogitto = { version = "6.5", default-features = false }
...
 indexmap = "2.10"
 serde = { version = "1.0", features = ["derive"] }
 inquire = "0.9"
+regex = "1.12"
+thiserror = "2"
 clap_complete_command = { version = "0.6", features = ["nushell"]}
 config = { version = "0.15", features = ["toml"] }
 xdg = "3.0"
 gix = "0.80.0"
+ast-grep-config = { version = "0.42.1" }
```
Comment (cococonscious) [inline Cargo.toml]: Please put these features behind a feature flag. I don't mind if it's in the default features, but I want to be able to exclude it if necessary.

File: Cargo.toml [hunk 4]
```diff
@@ -6,8 +6,8 @@ authors = [
   "Finley Thomalla <finley@thomalla.ch>",
   "Danny Tatom <its.danny@hey.com>",
 ]
-description = "An interactive CLI for creating conventional commits."
 documentation = "https://docs.rs/koji"
+description = "An interactive CLI for creating conventional commits."
```
Comment (cococonscious) [inline Cargo.toml:9]: What's up with this reordering?

File: src/lib/config.rs [hunk 5]
```diff
@@ -13,10 +14,15 @@ pub struct Config {
...
     pub breaking_changes: bool,
     pub commit_types: IndexMap<String, CommitType>,
+    pub commit_scopes: IndexMap<String, CommitScope>,
     pub emoji: bool,
     pub issues: bool,
     pub sign: bool,
+    pub force_scope: bool,
+    pub allow_empty_scope: bool,
     pub workdir: PathBuf,
+    pub scope_patterns: IndexMap<String, ScopePatternValue>,
```
Comment (cococonscious) [inline src/lib/config.rs]: I'm wondering if we should rather have the patterns be a property in the commit_scopes entries, it seems more fitting, otherwise you configure the same scope in multiple places.

File: src/lib/config.rs [hunk 6]
```diff
@@ -13,10 +14,15 @@ pub struct Config {
...
+    pub scope_ast_grep: Vec<ScopeAstGrepRule>,
```
Comment (cococonscious) [inline src/lib/config.rs]: Probably the same here, should maybe be a property on the commit_scopes entries.

File: src/lib/config.rs [hunk 7]
```diff
@@ -13,10 +14,15 @@ pub struct Config {
     pub autocomplete: bool,
     pub breaking_changes: bool,
     pub commit_types: IndexMap<String, CommitType>,
+    pub commit_scopes: IndexMap<String, CommitScope>,
     pub emoji: bool,
     pub issues: bool,
     pub sign: bool,
+    pub force_scope: bool,
```
Comment (cococonscious) [inline src/lib/config.rs]: I'm afraid I will have to insist on renaming this to `force_config_scopes` or similar. It must be clear that the scopes defined in the config are being forced, not that it's similar to `allow_empty_scope`.

File: README.md [hunk 8]
```diff
@@ -191,3 +191,27 @@ emoji = true
 issues = true
 ```
 
+#### `scope-patterns`
```
Comment (cococonscious) [inline README.md]: Make sure to document all new configuration parameters.

File: README.md [hunk 9]
```diff
@@ -191,3 +191,27 @@ emoji = true
...
+flakes = "/flake\\.nix$"
+core = "/crates/core/**/*.rs"
+build = ["^/build\\.rs$", "/justfile"]
+```
+
+#### Ast-grep Scope Config
+
+- Location: In the normal Koji config file alongside the rest of the scope config
+- Optional: `true`
+- Description: Ast-grep rules can pre-assign commit scopes from structural code matches.
```
Comment (cococonscious) [inline README.md]: I think it'd be great if the pre-assigned scope, when an ast-grep rule matches, would show up in the prompt somehow, I think nicest would be something like this:
```sh
> What type of change are you committing? feat
> What's the scope of this change? foo      <---- This line should have been printed automatically instead of being omitted.
> Write a short, imperative tense description of the change: bar
```

If this is not possible, for example if the prompting library doesn't allow for it, print it in some other fitting way, I just don't like that it's invisible.

File: src/bin/main.rs [hunk 10]
```diff
@@ -193,10 +194,13 @@ fn main() -> Result<()> {
         sign,
         _user_config_path: None,
         _current_dir: Some(current_dir.clone()),
+        ..Default::default()
     }))?;
 
+    let scope_matches = detect_scope_matches(&repo, &config)?;
```
Comment (cococonscious) [inline src/bin/main.rs:206]: If the `all` parameter is passed, newly staged files (that were previously not staged) will not be recognized.

File: src/lib/scope.rs [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    let mut matched_scopes = IndexSet::new();
+
+    for relative_path in changed_paths {
+        let applicable_rules = compiled_rules.rules.for_path(relative_path);
+        if applicable_rules.is_empty() {
+            continue;
+        }
+
+        let full_path = workdir.join(relative_path);
+        let Ok(source) = std::fs::read_to_string(&full_path) else {
```
Comment (cococonscious) [inline src/lib/scope.rs]: Correct me if I'm wrong but the way it seems to me is that it checks files by path, not by the staged blobs, meaning that if you make changes, stage them, then make changes again but don't stage those, the rules will match on the latter, unchanged changes as well.

Comment (cococonscious) [reply src/lib/scope.rs]: Make sure to add a test to catch this edge-case after fixing it.

File: src/lib/scope.rs [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+
+    Ok(())
+}
+
+pub fn validate_ast_grep_rules(config: &Config) -> Result<()> {
+    compile_ast_grep_rules(config).map(|_| ())
+}
+
+pub fn detect_scope_matches(repo: &Repository, config: &Config) -> Result<ScopeMatches> {
+    let changed_paths = staged_paths(repo)?;
```
Comment (cococonscious) [inline src/lib/scope.rs]: Is the behavior of this example expected?:

1. tests/a.rs renamed to src/a.rs  
2. old path (tests/a.rs) matches filter but cannot be read -> skipped  
3. new path (src/a.rs) can be read, but doesn’t match filter -> no rule applied  
=> result: no AST scope match, even though the staged change involved a test file rename.

**Review by cococonscious — CHANGES_REQUESTED** [review]
I like the general idea, but please do look at my review comments and answer / fix the issues, thanks! Please also don't forget to add unit tests for the new functions, covering both positive and negative cases, for example: `looks_like_regex`

PS: Sorry for the late replies and reviews 😮‍💨

File: Cargo.toml (hunk 4 above)
Comment (philocalyst) [reply Cargo.toml:9]: I was getting dependency conflicts and threw $LLM at it while I was finishing up the rest, likely an artifact from that hail mary. Will remove.

**Comment by philocalyst** [comment]
My own turn for a slow response! Pretty happy with how things are looking right now, wondering if you'd like to try to use it to add a few local rules for Koji itself and see if it's intuitive or not with a real use-case? That could be fun :)

File: README.md (hunk 9 above)
Comment (cococonscious) [reply README.md]: @philocalyst This is not resolved, it's still not showing the assigned scope:
```sh
> What type of change are you committing? feat
> Write a short, imperative tense description of the change: foo
> ...
feat(lib): foo
```

File: src/lib/scope.rs (hunk 11 above)
Comment (cococonscious) [reply src/lib/scope.rs]: @philocalyst Same here, not resolved.

File: Cargo.toml (hunk 3 above)
Comment (cococonscious) [reply Cargo.toml]: @philocalyst Also not resolved, compile with `--no-default-features` to check.

File: src/lib/scope.rs (hunk 12 above)
Comment (cococonscious) [reply src/lib/scope.rs]: @philocalyst I think this is also still open, if not, please answer with an explanation before resolving.

**Comment by philocalyst** [comment]
OKAY i think finally solid @cococonscious 

i was lowkey intimidated by my own feature so was avoiding testing it directly, my fault for the annoying back and forth

**Comment by philocalyst** [comment]
Seems the failing test is due to a missing key @cococonscious

**Comment by philocalyst** [comment]
Following up here!

**Review by cococonscious — DISMISSED** [review]
LGTM 🥇 
I'll need you to update the PR (rebase or merge, doesn't matter to me as I will squash it anyways, therefore a PR description would also be nice 👍 ) and upgrade the dependencies you added to their latest versions, after that I'll happily merge it. Thanks a lot for your contribution and sorry for the wait times.

**Comment by philocalyst** [comment]
Should be solid :)

**Review by cococonscious — APPROVED** [review]
